tests/integration/targets/networkfirewall/tasks/complex.yml

```
@@ -0,0 +1,612 @@
+        policy_change_protection: False
+        subnet_change_protection: False
+        subnets:
+        - '{{ subnet_id_a_1a }}'
+        - '{{ subnet_id_a_2 }}'
+        - '{{ subnet_id_a_3 }}'
+        tags: '{{ second_tags }}'
+      register: complex_firewall
+
+    - assert:
+        that:
+          - complex_firewall is changed
+          - '"firewall" in complex_firewall'
+          - '"firewall" in complex_firewall.firewall'
+          - '"delete_protection" in firewall_data'
+          - '"description" in firewall_data'
+          - '"firewall_arn" in firewall_data'
+          - '"firewall_id" in firewall_data'
+          - '"firewall_metadata" in complex_firewall.firewall'
+          - '"firewall_name" in firewall_data'
+          - '"firewall_policy_arn" in firewall_data'
+          - '"firewall_policy_change_protection" in firewall_data'
+          - '"subnet_change_protection" in firewall_data'
+          - '"subnet_mappings" in firewall_data'
+          - '"tags" in firewall_data'
+          - '"vpc_id" in firewall_data'
+          - firewall_data.delete_protection == true
+          - firewall_data.description == 'An updated Description'
+          - firewall_data.firewall_arn == complex_firewall_arn
+          - firewall_data.firewall_id == complex_firewall_id
+          - firewall_data.firewall_name == complex_firewall_name
+          - firewall_data.firewall_policy_arn == default_policy_arns[2]
+          - firewall_data.firewall_policy_change_protection == false
+          - firewall_data.subnet_change_protection == false
+          - firewall_data.subnet_mappings | length == 3
+          - '"subnet_id" in firewall_data.subnet_mappings[0]'
+          - subnet_id_a_1a in subnet_ids
+          - subnet_id_a_2 in subnet_ids
+          - subnet_id_a_3 in subnet_ids
+          - firewall_data.vpc_id == vpc_id_a
+          - firewall_data.tags == second_tags
+          - '"configuration_sync_state_summary" in firewall_metadata'
+          - '"status" in firewall_metadata'
+          - '"sync_states" in firewall_metadata'
+          - firewall_metadata.status == "READY"
+          - firewall_metadata.configuration_sync_state_summary == "IN_SYNC"
+      vars:
+        firewall_data: '{{ complex_firewall.firewall.firewall }}'
+        firewall_metadata: '{{ complex_firewall.firewall.firewall_metadata }}'
+        subnet_ids: '{{ firewall_data.subnet_mappings | map(attribute="subnet_id") | list }}'
+
+    # There's a delay of a couple of seconds between requesting a change and the
+    # firewall entering a PENDING/PROVISIONING state.  The delay is enough to
+    # let the firewall enter the expected PENDING state is it's going to.
+    - name: 'Pause to let synchroization start'
+      pause:
+        seconds: 5
+
+    # Also checks that we get the info we expect
+    - name: 'Check firewall is not still updating'
+      networkfirewall_info:
+        arn: '{{ complex_firewall_arn }}'
+      register: complex_firewall_info
```

```suggestion
    # Also checks that we get the info we expect
    - name: 'Check firewall is not still updating'
      networkfirewall_info:
        arn: '{{ complex_firewall_arn }}'
      register: complex_firewall_info
      retries: 5
      delay: 5
      until: "firewall_metadata.status == 'READY'"
```

what do you think of? Because it might happen that even 5s are not enough.

---

I possibly need to change the comment.  What's weird with Firewall is that you can make they change, it'll be reflected in the main firewall information, then, after a few seconds the firewall will enter the provisioning / synchronising states.

The default is to wait.  As such everything should be complete once we're done.

What this pause is intended to catch is when the wait **should** have happened, but has **not** happened.  By delaying 5 seconds, if it's going to enter the provisioning state after returning, then then the status should have left ready and entered provisioning:

Order of events
1) API requests change
2) API returns success
2a) Resource is **still** in ready state.
3) Delay (about 2/3 seconds from experience)
4) Resource enters "provisioning" state
5) Change takes up to 15 minutes to complete
6) Change completes
7) Resource enters "ready" state

---

Normally I like the "retry" approach, but with networkfirewall we're looking at at least 5-10 minutes for subnet/policy changes to propagate, so shaving 5 seconds from the test doesn't make much of a difference.